### DISCUSSION TITLE
Additional QR Code Customization Options

### DISCUSSION THREAD

**Discussion by ghost** [Shane32/QRCoder#454] [Q&A] [answered]
I'm playing around with this library and trying to do some "pretty" QR codes for use on things like business cards. I used one of those free online services to create a QR code to link to my Discord server that works and has most of the "features" I want - a logo in the center, "dots" instead of "blocks" for the encoded data, and "custom" position finder images. However, I really kinda want to do something myself that isn't tied to an online service.

I found that the `ArtQrCode` will give me the dots by default and will let me set the position finders but not add the logo/icon Other renderers (`SvgQRCode` and `Base64QRCode`) can add the logo/icon, but won't let me customize the other elements. I'm mainly curious if there's a reason that these specific options aren't "consistent" from one to the next.

I realize that I can do things like generate an `ArtQRCode` with the dots and position finders, then draw back over it with the logo although, since I'm new to generating my own QR code, I'm not 100% sure about the integrity of the QR code when doing that. I was just wondering if there's something I'm missing in the existing library that would allow me to "one-shot" all of that.

**Comment by codebude** [comment, marked as answer]
Currently there is no "one-shot" solution for your requirement.

>  I'm mainly curious if there's a reason that these specific options aren't "consistent" from one to the next.

That's because QRCoder grew over time and some renderers (like ArtQRCode) were coded by contributors and I weren't that strict that time, when it comes to consistency. (These days I would do it better.)

>  I'm not 100% sure about the integrity of the QR code when doing that.

The ECC (error correction code) level defines how much redundancy is generated into the QR code. The higher the ECC level, the more of QR code can be covered while maintaining it's readability at the same time. If you choose ECC level H (the highest) up to 30% of the QR code can be covered. (With some restrictions like: do not cover all finder patterns, etc.)
So rule of thumb: Choose ECC H  and ensure, that your logo doesn't cover more than 25% of all pixels of the QR code.
More on ECC levels: https://en.wikipedia.org/wiki/QR_code#Error_correction

**Reply by ghost** [reply]
Thank you so much for the response. 
> That's because QRCoder grew over time and some renderers (like ArtQRCode) were coded by contributors and I weren't that strict that time, when it comes to consistency. (These days I would do it better.)

That makes total sense (especially since I'm in the same boat with a bunch of the code I'm maintaining for my "day job").
> If you choose ECC level H (the highest) up to 30% of the QR code can be covered. (With some restrictions like: do not cover all finder patterns, etc.)

I *thought* that the ECC would be a big part of it, but, like I said, I haven't really tried to do this myself before, so I didn't want to make too many assumptions.

Thank you again for the help. I'll start playing around with this a bit more and see what I can do. Maybe I'll look into playing around in the source code a bit more to make them all function similarly for my own purposes.
